### PR TITLE
Updates to glib logging setup

### PR DESCRIPTION
**Pull request by clumens**

### CODE REVIEW

File: lib/common/logging.c [hunk 1]
```diff
@@ -830,11 +832,19 @@ crm_log_preinit(const char *entity, int argc, char *const *argv)
...
-    glib_log_default = g_log_set_default_handler(crm_glib_handler, NULL);
-
-    /* and for good measure... - this enum is a bit field (!) */
-    g_log_set_always_fatal((GLogLevelFlags) 0); /*value out of range */
+    /* Add a log handler for messages from our log domain at any log level.
+     * Any other log messages will go to whatever the default is.
+     */
+    g_log_set_handler(G_LOG_DOMAIN,
+                      G_LOG_LEVEL_MASK | G_LOG_FLAG_FATAL | G_LOG_FLAG_RECURSION,
+                      crm_glib_handler, NULL);
```
Comment (kgaillot) [inline lib/common/logging.c:840]: pcmk__log_id =

File: lib/common/logging.c [hunk 2]
```diff
@@ -47,7 +47,7 @@ unsigned int crm_trace_nonlog = 0;
 bool pcmk__is_daemon = false;
 
 static unsigned int crm_log_priority = LOG_NOTICE;
-static GLogFunc glib_log_default = NULL;
+static guint pcmk__log_id = 0;
```
Comment (kgaillot) [inline lib/common/logging.c:50]: Are you sure `g_log_set_handler()` can't return 0 as a valid ID? Hopefully glib's own default handler is 0

**Review by kgaillot — COMMENTED** [review]
nice

have you tried triggering a glib message to make sure it works?

File: lib/common/logging.c (hunk 2 above)
Comment (clumens) [reply lib/common/logging.c:50]: I don't think it actually uses IDs for the default handler - at least, not on main.  Looking at the source, it maintains a separate default_log_handler variable that is just assigned a function, not a handler object that would have an ID.

**Comment by clumens** [comment]
> have you tried triggering a glib message to make sure it works?

The g_warning/info/error messages work as expected.  Assertions are a different story.

For example, add something obviously wrong to crm_mon.c - like, `g_hash_table_remove_all(NULL)`.  With the code as it is in this PR, the assertion will be logged like so:

```
(crm_mon:3557947): GLib-CRITICAL **: 11:47:49.187: g_hash_table_remove_all: assertion 'hash_table != NULL' failed
```

Looking at the source, this uses `g_return_if_fail` which is a macro defined in gmessages.h which in turn calls `g_return_if_fail_warning`.  The macro uses G_LOG_DOMAIN, which we're setting to Pacemaker.  But wait - because this is defined in a header that gets included in ghash.c (where g_hash_table_remove_all is), it will use whatever G_LOG_DOMAIN is set to when glib itself is built.  Sure enough, if I check this with gdb, I see that it's using the "GLib" domain here.

So long story short, to catch assertions coming from within GLib itself, we also need to add a g_log_set_handler block for that domain too.  If we do that, we'll be redirecting any other GLib assertions to the pacemaker log that would come from other users of GLib + Pacemaker.

**Comment by kgaillot** [comment]
> > have you tried triggering a glib message to make sure it works?
> 
> The g_warning/info/error messages work as expected. Assertions are a different story.
> 
> For example, add something obviously wrong to crm_mon.c - like, `g_hash_table_remove_all(NULL)`. With the code as it is in this PR, the assertion will be logged like so:
> 
> ```
> (crm_mon:3557947): GLib-CRITICAL **: 11:47:49.187: g_hash_table_remove_all: assertion 'hash_table != NULL' failed
> ```
> 
> Looking at the source, this uses `g_return_if_fail` which is a macro defined in gmessages.h which in turn calls `g_return_if_fail_warning`. The macro uses G_LOG_DOMAIN, which we're setting to Pacemaker. But wait - because this is defined in a header that gets included in ghash.c (where g_hash_table_remove_all is), it will use whatever G_LOG_DOMAIN is set to when glib itself is built. Sure enough, if I check this with gdb, I see that it's using the "GLib" domain here.
> 
> So long story short, to catch assertions coming from within GLib itself, we also need to add a g_log_set_handler block for that domain too. If we do that, we'll be redirecting any other GLib assertions to the pacemaker log that would come from other users of GLib + Pacemaker.

Yeah we definitely need to catch GLib, that's the main goal (I don't think we use glib logging functions ourselves). I don't think that stops an external library user from overriding that with their own GLib handler though.

I wonder whether there are any other domains glib might trigger indirectly, like for gio.

**Comment by clumens** [comment]
Updated, though I haven't investigated the GIO case yet.  I'm also unsure if I should add `g_log_set_fatal_mask("GLib", 0);` as well.

**Comment by clumens** [comment]
Well, the bad news is that yes, GLib does have additional domains - "GLib-GIO" is the one we are most likely to run into, but there's a handful of others.  And log domain matching just does strcmp.  It doesn't do anything smarter with whatever comes after the dash.  I can definitely add others, but I wonder if we'll just always be one step behind whatever glib does.

**Review by kgaillot — COMMENTED** [review]
s/include/internal/ in commit message

libgio is in the glib source repo, so hopefully it uses the same log domain

**Comment by kgaillot** [comment]
> Well, the bad news is that yes, GLib does have additional domains - "GLib-GIO" is the one we are most likely to run into, but there's a handful of others. And log domain matching just does strcmp. It doesn't do anything smarter with whatever comes after the dash. I can definitely add others, but I wonder if we'll just always be one step behind whatever glib does.

Bah. This whole interface seems not well thought out.

Oddly I can't find where I saw that libraries shouldn't set a default handler, though it makes sense. The code does say libraries shouldn't set the global fatal mask.

I guess let's just catch the existing domains.

File: lib/common/logging.c [hunk 3]
```diff
@@ -129,9 +130,12 @@ crm_trigger_blackbox(int nsig)
...
 crm_log_deinit(void)
 {
-    if (glib_log_default != NULL) {
-        g_log_set_default_handler(glib_log_default, NULL);
+    if (pcmk__log_id == 0) {
+        return;
     }
+
+    g_log_remove_handler(G_LOG_DOMAIN, pcmk__log_id);
+    g_log_remove_handler("GLib", pcmk__glib_log_id);
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/common/logging.c:138]: I don't believe it can - looking at the definition of `g_log_set_handler_full` (https://gitlab.gnome.org/GNOME/glib/-/blob/main/glib/gmessages.c?ref_type=heads#L732) it returns 0 for the handler ID on error and it pre-increments the global `handler_id` variable that it hands out so it should always be starting with 1.

The fact that it returns 0 on error makes me think it can't possibly be used internally to identify the default handler.  Plus, the default handler is referenced by `default_log_func` throughout and not by ID as far as I can tell.

File: lib/common/logging.c [hunk 4]
```diff
@@ -129,9 +133,20 @@ crm_trigger_blackbox(int nsig)
...
+    g_log_remove_handler(G_LOG_DOMAIN, pcmk__log_id);
+    pcmk__log_id = 0;
+    g_log_remove_handler("GLib", pcmk__glib_log_id);
+    pcmk__glib_log_id = 0;
+    g_log_remove_handler("GLib-GIO", pcmk__gio_log_id);
+    pcmk__gio_log_id = 0;
+    g_log_remove_handler("GModule", pcmk__gmodule_log_id);
+    pcmk__gmodule_log_id = 0;
+    g_log_remove_handler("GThread", pcmk__gthread_log_id);
+    pcmk__gthread_log_id = 0;
```
Comment (kgaillot) [inline lib/common/logging.c:149]: We most likely will never use GModule or GThread, so we can drop those

**Review by kgaillot — COMMENTED** [review]
libcrmservice can't find crm_internal.h -- should be able to just include common.mk in lib/services

File: lib/common/logging.c (hunk 4 above)
Comment (clumens) [reply lib/common/logging.c:149]: They are both potentially used by glib somewhere - gthread looks like it could be used by mainloop stuff, and gmodule is potentially used by gio.

**Comment by clumens** [comment]
> libcrmservice can't find crm_internal.h -- should be able to just include common.mk in lib/services

Yeah, I just found that.  And then MAINTAINERCLEANFILES needs to be removed from `lib/services/Makefile.am`.  It's probably worth making a pass over all makefiles to check that they include common.mk.

File: lib/cib/Makefile.am [hunk 5]
```diff
@@ -21,7 +21,7 @@ libcib_la_SOURCES	+= cib_remote.c
 libcib_la_SOURCES	+= cib_utils.c
 
 libcib_la_LDFLAGS	= -version-info 33:0:6
-libcib_la_CPPFLAGS	= -I$(top_srcdir) $(AM_CPPFLAGS)
+libcib_la_CPPFLAGS 	= $(AM_CPPFLAGS)
```
Comment (kgaillot) [inline lib/cib/Makefile.am:24]: We don't need target_CPPFLAGS unless we need to add something to AM_CPPFLAGS (but that can go in the next PR)

**Review by kgaillot — APPROVED** [review]

File: lib/cib/Makefile.am (hunk 5 above)
Comment (clumens) [reply lib/cib/Makefile.am:24]: I thought that was the case too, but there was some weirdness going on with failures in CI.

**Comment by nrwahl2** [comment]
@clumens I wonder if this fixed an actual issue that was reported somewhere, or if it was a "just in case" thing. I like the idea of setting the default handler, like we used to do, plus setting a handler for `G_LOG_DOMAIN`. Trying to keep up with all of GLib's domains and set handlers for them explicitly is not ideal.

Currently we set handlers for "GLib", "GLib-GIO", "GModule", and "GThread". However, GLib also has at least the following other domains:
* "GLib-GirCompiler"
* "GLib-GirDecompiler"
* "GLib-GirInspector"
* "GIRepository"
* "GLib-GIRepository"
* "GLib-GObject" (added in 2016 by commit [21395797](https://gitlab.gnome.org/GNOME/glib/-/commit/21395797) in GLib 2.53)
* "GLib-GRegex" (added in 2016 by commit [fe2a9887](https://gitlab.gnome.org/GNOME/glib/-/commit/fe2a9887) in GLib 2.53, removed in 2021 by commit [9570e677](https://gitlab.gnome.org/GNOME/glib/-/commit/9570e677) in GLib 2.69)

I doubt we'll ever use the "Gir"/"GIRepository" ones, but still.

We're currently tied to GLib 2.42, but you can see how this becomes an issue for future-proofing.

If anything external links against Pacemaker, its log handler will be unaffected unless it calls `crm_log_preinit()`. To be fair, I think everything is supposed to call that, since we do general initialization there. However, see https://projects.clusterlabs.org/T840.

I dunno... I can see the motivation. If we rely on the default log handler and something else resets it, we're out of luck. Then again, something else could theoretically set a different log handler for the GLib domains too.

Another possibility would be to set handlers for only the parts of GLib that we use -- namely, `"GLib"` and `"GLib-GIO"`. This still presents a maintenance issue if we ever start using other sub-libraries of GLib, because we'd need to remember to add a handler.

**Comment by clumens** [comment]
> @clumens I wonder if this fixed an actual issue that was reported somewhere, or if it was a "just in case" thing. I like the idea of setting the default handler, like we used to do, plus setting a handler for `G_LOG_DOMAIN`. Trying to keep up with all of GLib's domains and set handlers for them explicitly is not ideal.

I don't remember exactly what was going on here, though I do have a vague memory of looking through the list of GLib domains.

I think I read somewhere that libraries shouldn't use the default GLib logging domain, and that we shouldn't override the default logging handler because that would also affect third-party programs linked with libpacemaker.  This led me down the path of introducing our own logging domain for pacemaker, but then once I'd done all that, we weren't getting glib messages in our log files which was making debugging harder.  So I added handlers for the glib-related domains.

Whether that's actually how anything still works, I dunno.  I can't remember looking at any of the logging related code since.

>     * "GLib-GObject" (added in 2016 by commit [21395797](https://gitlab.gnome.org/GNOME/glib/-/commit/21395797) in GLib 2.53)

This one could possibly be a concern in the future.

> We're currently tied to GLib 2.42, but you can see how this becomes an issue for future-proofing.

Yep.

> If anything external links against Pacemaker, its log handler will be unaffected unless it calls `crm_log_preinit()`. To be fair, I think everything is supposed to call that, since we do general initialization there. However, see https://projects.clusterlabs.org/T840.

Library/logging initialization (especially when it comes to being used by third party programs) seems like a big mess to me at the moment.

> I dunno... I can see the motivation. If we rely on the default log handler and something else resets it, we're out of luck. Then again, something else could theoretically set a different log handler for the GLib domains too.
> 
> Another possibility would be to set handlers for only the parts of GLib that we use -- namely, `"GLib"` and `"GLib-GIO"`. This still presents a maintenance issue if we ever start using other sub-libraries of GLib, because we'd need to remember to add a handler.

That's essentially what I was going for with that list.  I do not remember exactly why GModule and GThread are included, but I don't think I added them just for fun.  It's possible something we're using in turn uses something that logs in those domains.
